### PR TITLE
(CDAP-677) Feature/support header query param

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

File: src/test/java/co/cask/http/HttpServerTest.java [hunk 1]
```diff
@@ -372,6 +367,38 @@ public void testChunkResponse() throws IOException {
     }
   }
 
+  @Test
+  public void testStringQueryParam() throws IOException {
+    testContent("/test/v1/stringQueryParam/mypath?name=netty", "mypath:netty", HttpMethod.GET);
```
Comment (ParthGandhi) [inline src/test/java/co/cask/http/HttpServerTest.java:377]: We should add a test case that doesn't send a `QueryParam` when expected in the handler, and assert that it defaults to null.

File: src/main/java/co/cask/http/HttpResourceModel.java [hunk 2]
```diff
@@ -147,4 +151,97 @@ public String toString() {
...
+          parameterInfo = ParameterInfo.create(annotation,
+                                               ParamConvertUtils.createQueryParamConverter(parameterTypes[i]));
+        } else if (HeaderParam.class.isAssignableFrom(annotationType)) {
+          parameterInfo = ParameterInfo.create(annotation,
+                                               ParamConvertUtils.createHeaderParamConverter(parameterTypes[i]));
+        }
+        paramAnnotations.put(annotationType, parameterInfo);
+      }
+
+      // Must have either @PathParam, @QueryParam or @HeaderParam, but not two or more.
```
Comment (ParthGandhi) [inline src/main/java/co/cask/http/HttpResourceModel.java:238]: Is this part of the spec too? I can't seem to find it.

Comment (ParthGandhi) [inline src/main/java/co/cask/http/HttpResourceModel.java:238]: Ignore ^. I misunderstood what you were doing.

File: src/main/java/co/cask/http/ParamConvertUtils.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+/**
+ * Util class to convert request parameters.
+ */
+public final class ParamConvertUtils {
+
+  private static final Map<Class<?>, Method> PRIMITIVES_PARSE_METHODS;
+
+  // Setup methods for converting string into primitive/boxed types
+  static {
+    Map<Class<?>, Method> methods = Maps.newIdentityHashMap();
```
Comment (ParthGandhi) [inline src/main/java/co/cask/http/ParamConvertUtils.java:53]: Why need an `IdentityHashMap`? Won't the classes always be distinct anyway?

Comment (chtyim) [inline src/main/java/co/cask/http/ParamConvertUtils.java:53]: Yes, but I need a map from Boxed type to the "valueOf" method of that boxed type.

**Comment by ParthGandhi** [comment]
Think we should have 1 more test case.

Other than that, LGTM.

**Comment by chtyim** [comment]
Which one more test?

**Comment by ParthGandhi** [comment]
I'm referring to https://github.com/caskdata/netty-http/pull/9/files#r20623806

**Comment by chtyim** [comment]
@ParthGandhi I added some more test cases and also added support for `@DefaultValue` (because it's useful in many cases).

File: src/main/java/co/cask/http/HttpResourceModel.java [hunk 4]
```diff
@@ -147,4 +152,125 @@ public String toString() {
...
+          parameterInfo = ParameterInfo.create(annotation, null);
+        }
+
+        paramAnnotations.put(annotationType, parameterInfo);
+      }
+
+      // Must have either @PathParam, @QueryParam or @HeaderParam, but not two or more.
+      if (Sets.intersection(SUPPORTED_PARAM_ANNOTATIONS, paramAnnotations.keySet()).size() != 1) {
+        throw new IllegalArgumentException(
+          String.format("Missing have exactly one annotation from %s for parameter %d in method %s",
```
Comment (ParthGandhi) [inline src/main/java/co/cask/http/HttpResourceModel.java]: "Must have exactly one annotation..."

**Comment by ParthGandhi** [comment]
One minor comment on the exception message. 

Otherwise LGTM.

File: src/main/java/co/cask/http/HttpResourceModel.java (hunk 4 above)
Comment (chtyim) [inline src/main/java/co/cask/http/HttpResourceModel.java]: Fixed.
